File: LabRasp1/Lab 9/Part 4/calculator.cpp
#include "calculator.h"

Calculator::Calculator(std::string path)
{
	builder = new HtmlBuilder(path);
	builder->createHtml()->addHead()->startBody()->createTable();

	sizes = std::vector<int>();
}

void Calculator::addTask(int size)
{
	sizes.push_back(size);
}

void Calculator::run()
{
	double sequentialTime;
	double time, acceleration;

	for (auto i : sizes) {
		std::cout << "Calculating matrix size of " << i << " ..." << std::endl;
		std::vector<std::pair<double, double>> results = std::vector<std::pair<double, double>>();
		sequentialTime = calculate(i, 1);
		std::cout << "\t1 core\t" << sequentialTime << std::endl;

		time = calculate(i, 2);
		acceleration = sequentialTime / time;
		results.push_back(std::pair<double, double>(time, acceleration));
		std::cout << "\t2 cores\t" << time << std::endl;

		time = calculate(i, 4);
		acceleration = sequentialTime / time;
		results.push_back(std::pair<double, double>(time, acceleration));
		std::cout << "\t4 cores\t" << time << std::endl;

		builder->addResult(i, sequentialTime, results);
	}
}

double Calculator::calculate(int size, int threadsNumber)
{
	auto matrixGenerator = MatrixGenerator(size, 100.0);
	auto A = matrixGenerator.generate();
	auto B = matrixGenerator.generate();

	auto stripesShema = StripesDivider(A, B, size, size, threadsNumber);

	auto start = std::chrono::steady_clock::now();

	auto C = stripesShema.calculateProduct();

	auto end = std::chrono::steady_clock::now();
	std::chrono::duration<double> elapsedSeconds = end - start;

	return elapsedSeconds.count();
}

void Calculator::finish()
{
	builder->finishTable()->finishBody()->finishHtml();
}

File: LabRasp1/Lab 9/Part 1/src/main/java/Calculator.java
import java.util.LinkedList;
import java.util.List;

public class Calculator {
    private HTMLBuilder htmlBuilder;
    private List<Integer> sizes;

    Calculator() {
        sizes = new LinkedList<>();

        htmlBuilder = new HTMLBuilder();
        htmlBuilder.build();
        htmlBuilder.createTable();
    }

    void addTask(Integer size) {
        sizes.add(size);
    }

    void run() {
        double sequentialTime;
        double time, acceleration;
        List<Pair<Double, Double>> results;

        for (Integer i : sizes) {
            System.out.println("Calculating matrix size of " + i + " ...");
            results = new LinkedList<>();
            sequentialTime = ((double) calculate(i, 1)) / 1000.0;
            System.out.println(ANSI.Yellow("\t1 core\t") + sequentialTime);

            time = ((double) calculate(i, 2)) / 1000.0;
            acceleration = sequentialTime / time;
            results.add(new Pair<>(time, acceleration));
            System.out.println(ANSI.Yellow("\t2 cores\t") + time);


            time = ((double) calculate(i, 4)) / 1000.0;
            acceleration = sequentialTime / time;
            results.add(new Pair<>(time, acceleration));
            System.out.println(ANSI.Yellow("\t4 cores\t") + time + "\n");


            htmlBuilder.addResult(i, sequentialTime, results);
        }

        finish();
    }

    private Long calculate(int size, int threadsNumber) {
        MatrixGenerator matrixGenerator = new MatrixGenerator(size, 100);
        double[][] A = matrixGenerator.generate();
        double[][] B = matrixGenerator.generate();

        StripesDivider stripesSchema = new StripesDivider(A, B, threadsNumber);
        long startTime = System.currentTimeMillis();

        stripesSchema.calculateProduct();

        return System.currentTimeMillis() - startTime;
    }

    private void finish() {
        htmlBuilder.finishTable();
        htmlBuilder.finish();
    }
}
